만국박람회 [STEP 3] 수꿍, yeton

Jeon-Minsu opened the pull request:
안녕하세요, 쿠마 @leejun6694  🤗🤗

벌써 프로젝트의 마지막 스텝이네요!
오토레이아웃이 이번 프로젝트의 주요 논제였는데, 생각보다 많은 고민과 의문이 생겼습니다.
이에 프로젝트 구현 과정에서 고민했던 점과 의문점에 대해 명시해보았습니다.

## 고민했던 점

### 주제
### 1. 디바이스 회전 제한 메서드 호출 범위

- 프로젝트 요구사항으로 첫 화면인 포스터는 세로로만 볼 수 있도록 하고, 출품작 목록과 상세 두 화면은 모두 가로로 회전이 가능하도록 명시되어 있었다.
- `AppDelegate` 파일 내, 디바이스 회전을 구현하는 메서드를 작성하여, `ViewController`에서 아래의 코드를 호출하기만 하면, 디바이스 회전을 제한할 수 있었다.
- 코드
    ```swift
    private let appDelegate = UIApplication.shared.delegate as! AppDelegate

    appDelegate.shouldSupportAllOrientation = false
    ```
- 그런데 이 메서드를, 처음에는 다음과 같이 각각의 `ViewController`의 `viewDidLoad` 메서드 내에서, 호출해야 한다고 생각했다.
- 코드
    ```swift
    final class EXPOInformationViewController: UIViewController {
        private let appDelegate = UIApplication.shared.delegate as! AppDelegate

        override func viewDidLoad() {
            super.viewDidLoad()

            appDelegate.shouldSupportAllOrientation = false
    }


    final class KoreaEntryViewController: UIViewController {
        private let appDelegate = UIApplication.shared.delegate as! AppDelegate

        override func viewDidLoad() {
            super.viewDidLoad()

            appDelegate.shouldSupportAllOrientation = true
    }


    final class KoreaEntryDetailsViewController: UIViewController {
        private let appDelegate = UIApplication.shared.delegate as! AppDelegate

        override func viewDidLoad() {
            super.viewDidLoad()

            appDelegate.shouldSupportAllOrientation = true
    }
    ```
- 하지만, 이와 같이 진행을 한다면, `appDelegate` 상수 선언이 반복적으로 사용되어, 싱글톤과 같은 방법을 필요로 하고, 두번째와 세번째 뷰컨트롤러는 설정 값이 같기 때문에 중복적으로 호출된 느낌이 받았으며, 또한, `viewDidLoad`에서 호출시, 뷰가 메모리에 올라간 한번만 호출되기 때문에 지속적으로 디바이스 회전을 제한하지 못하였다.
- 위의 문제점을 종합해본 결과, 다음과 같이, 첫번째 뷰컨트롤러의 `viewWillAppear`에 세로 모드만 지원하고, `viewWillDisAppear`에서 이 제약을 다시 풀어지면, 두번째 및 세번째 뷰컨트롤러에서는 화면 회전이 정상적으로 작동하였고, `viewDidLoad`에서 구현하지 않아 지속적으로 디바이스 회전을 제한하였으며, `appDelegate` 상수 선언을 최초 1회만 실행하면 되었다.
- 코드(`EXPOInformationViewController`)
    ```swift
    final class EXPOInformationViewController: UIViewController {
        // MARK: Properties

        ...

        private let appDelegate = UIApplication.shared.delegate as! AppDelegate

        ...

        // MARK: Life Cycle

        ...

        override func viewWillAppear(_ animated: Bool) {
            super.viewWillAppear(animated)

            appDelegate.shouldSupportAllOrientation = false
            ...
        }

        override func viewWillDisappear(_ animated: Bool) {
            super.viewWillDisappear(animated)

            appDelegate.shouldSupportAllOrientation = true
            ...
        }
    }
    ```

### 2. ContentConfiguration 이미지 크기 제한

- `ContentConfiguration` 기능을 통하여, 셀 내 이미지의 크기를 제한하고 싶어 다음과 같이 코드를 작성하였으나, 이미지와 텍스트의 정렬이 이미지 크기의 차이로 인하여 깔끔하게 정렬이 이루어지지 않았다.
- 코드
```swift
extension KoreaEntryViewController: UITableViewDataSource {
    ...
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: NameSpace.entryCellId.name, for: indexPath)
        var content = cell.defaultContentConfiguration()
        
        tableView.separatorInset.left = .zero
                
        content.text = koreaEntry[indexPath.row].name
        content.secondaryText = koreaEntry[indexPath.row].shortDescription
        content.image = UIImage(named: koreaEntry[indexPath.row].imageName)

        content.textProperties.font = UIFont.preferredFont(forTextStyle: .title2)
        content.secondaryTextProperties.font = UIFont.preferredFont(forTextStyle: .body)
        content.imageProperties.maximumSize.width = 50
        content.imageProperties.maximumSize.height = 50
                
        cell.contentConfiguration = content
        cell.accessoryType = .disclosureIndicator

        return cell
    }
}
```
- 예시 사진
<img width="340" alt="스크린샷 2022-06-23 오후 8 53 58" src="https://user-images.githubusercontent.com/99063327/175292809-f38df45a-c0d8-44ab-b28f-c7549fed4a20.png">
- 이를 해결하기 위해선, 각각의 이미지 크기에 상관없이, 이미지의 레이아웃 크기를 설정해줄 필요가 있었다.
- 학습 결과, `reservedLayoutSize`를 통해 이미지의 레이아웃 크기를 설정함으로써, 이미지의 크기에 관계없이 이미지와 텍스트 간 정렬이 올바르게 적용되었다.
- 예시 사진
<img width="337" alt="스크린샷 2022-06-23 오후 9 01 48" src="https://user-images.githubusercontent.com/99063327/175294079-e7a36a38-d575-4c21-af70-894ac7871250.png">
- 주의할 점은, `reservedLayoutSize`는 이미지의 크기를 결정하지 않으므로, `maximumSize`와 함께 사용하기를 권한다.


### 3. 한 string 내에서 폰트의 크기가 달라져야 할 경우 
- 명세를 보니, `"방문객 수 : \(visitors) 명"` 라는 `string` 내에서 `방문객 수 :`와 `\(visitors)명`의 폰트 크기가 다른 것을 확인하였습니다. 
- 그래서 하나의 문자열 내에서 특정 부분만 font를 바꿔주는 것에 대해 두 가지의 방법을 생각해보았습니다.

[첫 번째 방법]
- `label`을 아예 2개 만든 후 첫 번째 `label`에 `방문객 수 :`을 넣고, 두 번째 `label`에 `\(visitors)명`을 넣은 뒤 첫 번째 `label`의 `font`를 지정해주는 방법

[두 번째 방법]
- `label`은 1개를 만든 후, 그 안에서 처리하는 방법
- `UILabel`을 `extension`해서 바꾸고자 하는 `string` 부분의 `range`를 구한 뒤 `NSMutableAttributedString`을 사용하여 그 범위 부분만 바꿔주는 방법

- 코드
```swift
extension UILabel {
    func changeFontSize(font: UIFont, targetString: String) {
        let fullText = text ?? ""
        let range = (fullText as NSString).range(of: targetString)
        let attributedString = NSMutableAttributedString(string: fullText)
        attributedString.addAttribute(.font, value: font, range: range)
        attributedText = attributedString
    }
}

```
- 이번 프로젝트에서는 라벨을 한 개만 생성하여 `UILabel`을 `extension` 해주는 두 번째 방법을 사용하여 해결하였습니다.



## 의문점

### 주제
### 1. UIApplication.shared.delegate as! AppDelegate 와  AppDelegate()의 차이

- `AppDelegate` 파일 내 클래스에서, 디바이스의 회전을 제한하는 메서드를 다음과 같이 구현하였습니다.
- 코드(`AppDelegate`)
    ```swift
    final class AppDelegate: UIResponder, UIApplicationDelegate {
        // MARK: Properties

        var shouldSupportAllOrientation = true

        // MARK: UI

        func application(_ application: UIApplication, supportedInterfaceOrientationsFor window: UIWindow?) -> UIInterfaceOrientationMask {
            if shouldSupportAllOrientation == true {
                return UIInterfaceOrientationMask.all
            }

            return UIInterfaceOrientationMask.portrait
        }
        ...
    }
    ```
- 코드(`EXPOInformationViewController`)
    ```swift
    final class EXPOInformationViewController: UIViewController {
        // MARK: Properties

        ...

        private let appDelegate = UIApplication.shared.delegate as! AppDelegate

        ...

        // MARK: Life Cycle

        ...

        override func viewWillAppear(_ animated: Bool) {
            super.viewWillAppear(animated)

            appDelegate.shouldSupportAllOrientation = false
            ...
        }

        override func viewWillDisappear(_ animated: Bool) {
            super.viewWillDisappear(animated)

            appDelegate.shouldSupportAllOrientation = true
            ...
        }
    }
    ```
- 이에 `EXPOInformationViewController`에서 `appDelegate`이란 상수를 선언함에 있어, `UIApplication.shared.delegate as! AppDelegate` 로 값을 할당해주었습니다.
- 상수를 선언함에 있어 `AppDelegate()`로 상수의 값을 할당해주면 안되는 것인지 의문이 제기되었습니다.
- 이를 고민한 결과, `UIApplication.shared.delegate`는 현재 구현한 만국박람회 어플리케이션의 `delegate`를 `AppDelegate`로 다운캐스팅을 한 것이므로, 시뮬레이터 내 디바이스 회전에 대한 메서드가 정상적으로 가동됩니다.
- 하지만, 그저 `AppDelegate()`로 설정을 하면, 만국박람회의 어플리케이션의 `AppDelegate`가 아닌, 일반적인 `AppDelegate` 타입이 생성된 것임을 깨달았습니다.
- 이를 비유하자면, 모나미 볼펜을 구매하기 위해서는, 모나미 볼펜을 생성하는 공장(`UIApplication`)에서 생산한 모나미 볼펜(`UIApplication.shared.delegate`)을 구매하여야 하는것이지, `AppDelegate()`는 어디서 생산된지도 모르고, 모나미 볼펜도 아닌, 그냥 볼펜을 산 것과 같은 의미로 이해해 보았습니다.
- 이와 같이, 이해해 본 바가 올바른 이해인지 여쭤보고 싶습니다!

### 2. 이미지도 dynamic type이 적용되야 하는지
- 첫 번째 `viewController`의 태극기 이미지와 두 번째 `viewController`의 테이블 뷰 왼쪽 이미지를 `accessibility Inspector`로 키워도 이미지 크기가 커지지 않았습니다.
- 그 이유는 첫 번째 `viewController`의 태극기 이미지는 1:1 비율로 크기를 정해주었기 때문이고, 두 번째 `viewController`의 이미지는 `width` 크기를 지정해주었기 때문입니다.

[두 번째 `viewController`의 이미지 부분 오토레이아웃] 
```swift
 entryImageView.centerYAnchor.constraint(equalTo: contentView.centerYAnchor).isActive = true
        entryImageView.leadingAnchor.constraint(equalTo: contentView.leadingAnchor, constant: 5.0).isActive = true
        entryImageView.widthAnchor.constraint(equalTo: contentView.widthAnchor, multiplier: 0.2).isActive = true
        entryImageView.heightAnchor.constraint(equalTo: heightAnchor).isActive = true
```
- 그래서 혹시 두 개의 이미지 크기도 다이나믹 하게 적용이 되야 하는지 쿠마에게 여쭤보고 싶습니다. 만일 적용이 되야 한다면 텍스트도 함께 커지게 될 때 이미지가 어느정도의 범위까지 커져야 하는지도 궁금합니다..!

### 3. ContentConfiguration과 Custom Cell

- `UITableViewDataSource`의 메소드인 `func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath)`에서 이미지 및 텍스트를 추가하기 위한 방법으로 두 가지 방법을 발견하였다. - 하나는, `cell.imageView.image` 나 `cell.textLabel.text`로 접근하는 방식이었고, 다른 하나는 `ContentConfiguration`을 통해 접근하는 방식이었다.
- 전자의 방법은 iOS 14를 끝으로 `deprecated` 되는 방법이기 때문에, 후자의 방법을 사용하는 것이 좋다고 판단하였다.
- 그런데, `ContentConfiguration`을 사용던 중, 상수 값이 아닌 다음과 같이 `image`의 `maximumSize`와 `reservedLayoutSize`를 `cell`의 `frame`을 기준으로 만들어보니, 스크롤을 반복할수록, 셀의 크기가 비정상적으로 증가하였다.
- 코드
    ```swift
    content.imageProperties.maximumSize.width = cell.frame.width * 0.1
    content.imageProperties.maximumSize.height = cell.frame.height
    content.imageProperties.reservedLayoutSize.width = cell.frame.width * 0.1
    content.imageProperties.reservedLayoutSize.height = cell.frame.height
    ```
- 예시 영상
![셀_재사용_문제](https://user-images.githubusercontent.com/99063327/175296689-e4d60806-aa70-4e02-a411-9c5646b358c6.gif)
- 이 문제는 셀의 재사용 때문에 발생하는 것으로 추정되는데, 이를 해결하기 위해서는 `prepareForReuse` 메서드를 통해 셀의 높이를 초기화해야 된다고 생각한다.
- 이를 사용하기 위해서는 `custom cell`을 생성하여 `UITableViewCell` 클래스를 상속하여야 하므로, `custom cell`을 생성하지 않고는 해결하기 어렵다고 판단하였다.
- 이를 통해, 두 가지 질문을 드리고 싶습니다.
- 하나는 `contentConfiguration` 내에서 셀 재사용 문제를 해결할 방법이 없는지 궁금합니다.
- 다음으로, `Cell.frame.height`가 저희가 생각하는 셀의 크기를 의미하는 것인지 여쭈어보고 싶습니다.

### 4. 오토레이아웃 문제


### [문제가 발생한 Cell 구조]
<img src = "https://user-images.githubusercontent.com/98514397/175303717-2bddc656-a91a-455b-b44a-96d74709ef3e.png" width = "70%" height = "10%"> 

### [첫 번째 문제]

- 위와 같은 `cell`에서 `imageView`의 오토레이아웃을 잡아주는 데에 있어 문제가 있었습니다.
- 맨 처음에는 아래와 같이 `imageView`의 오토레이아웃을 잡아주었는데, 아래처럼 `contentView`에 `heightAnchor`를 주면 높이가 한없이 커지는 현상이 발생했습니다. 

```swift
class EntryTableViewCell {
    entryImageView.centerYAnchor.constraint(equalTo:     contentView.centerYAnchor).isActive = true
        entryImageView.leadingAnchor.constraint(equalTo: contentView.leadingAnchor, constant: 5.0).isActive = true
        entryImageView.widthAnchor.constraint(equalTo: contentView.widthAnchor, multiplier: 0.2).isActive = true
        entryImageView.heightAnchor.constraint(equalTo: contentView.heightAnchor).isActive = true
}
```

- 그래서 아래 코드와 같이 `entryImageView`의 `heightAnchor`를 `contentView`가 아닌 `Cell` 자체의 `heightAnchor`에 지정해주었더니 정상적으로 셀이 보였습니다.
```swift 
entryImageView.heightAnchor.constraint(equalTo: heightAnchor).isActive = true
```

- 둘의 차이는 `imageView`의 `height`를 준 기준을 `contentView`로 준 것과 `cell`로 준 차이밖에 없는데 이렇게 다른 결과가 나오는 이유가 무엇인지 궁금합니다. Cell의 `ContentView`에 정해지지 않은 이미지의 크기를 주면 `ContentView`의 크기도 같이 커져버리는 것인지.. 정확히 이해가 가지 않습니다. 쿠마의 조언이 필요합니다!


### [두 번째 문제]
 
```swift
class EntryTableViewCell {
    private func setAutoLayout() {
       stackView.addArrangedSubview(titleLabel)
       stackView.addArrangedSubview(descriptionLabel)
       contentView.addSubview(entryImageView)
       contentView.addSubview(stackView)
        
        entryImageView.topAnchor.constraint(equalTo: topAnchor).isActive = true
        entryImageView.bottomAnchor.constraint(equalTo: bottomAnchor).isActive = true
        entryImageView.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 10.0).isActive = true
        entryImageView.trailingAnchor.constraint(equalTo: stackView.leadingAnchor, constant: -10.0).isActive = true
        entryImageView.widthAnchor.constraint(equalTo: widthAnchor, multiplier: 0.2).isActive = true
        
        
        stackView.topAnchor.constraint(equalTo: contentView.topAnchor, constant: 10).isActive = true
        stackView.trailingAnchor.constraint(equalTo: contentView.trailingAnchor, constant: -10.0).isActive = true
        stackView.bottomAnchor.constraint(equalTo: contentView.bottomAnchor, constant: -10).isActive = true
    }
}
```
- `EntryTableCell` 내부 이미지 뷰와 스택 뷰의 오토레이아웃을 위와 같이 잡아주었습니다.
- 위와 같이 오토레이아웃을 잡으니 `numberOfLines = 0`을 적용해주어도 화면에 stackView 내부의 descriptionLabel의 전체가 나타나지 않고 `...`처럼 텍스트가 잘리는 현상이 발생했습니다.
- `top`과 `bottom`을 잡아주어서 그런건지.. 왜 이렇게 나오는지 이유를 모르겠어서 쿠마에게 여쭙고 싶습니다.

[오류가 나는 화면]

 <img src = "https://user-images.githubusercontent.com/98514397/175303548-f6de696a-febf-4cf2-a447-6b8d24b8e9a4.png" width = "30%" height = "10%"> 

### 5. Navigation Back Button Title 변경 문제

- `Accessibility`를 통해 `Dynamic Type`의 크기를 조절해본 결과, 네비게이션 바의 `back button title`의 크기가 `navigation bar title`의 영역을 침범하게 되면, `back button title`이 `back`으로 변경되는 것을 확인하였습니다.
- 예시 영상
![백버튼_타이틀_변경](https://user-images.githubusercontent.com/99063327/175301552-4195d4ee-7167-47c9-a246-7008301bf8e1.gif)
- 이에 다이나믹 타입을 사용하면서 본래의 `back button title`을 유지하기 위한 방법이 있는지 궁금합니다.


---

이상 PR 마치겠습니다.
이번 프로젝트 동안 좋은 피드백 해주셔서 정말 감사했습니다!!
마지막 step3도 잘 부탁드립니다!!
나중에 또 뵈면 좋겠습니다 🤗🤗

Jeon-Minsu committed:
refactor: 불필요한 NameSpace 제거

- File name, Identifier와 같이 구현에 crash를 발생시키는 case를 제외한 불필요한 NameSpace 제거

yeeton37 committed:
feat: 첫번째 화면 가로모드 안되도록 설정

yeeton37 committed:
Merge branch 'step2' into step3

Jeon-Minsu committed:
refactor: 코딩 컨벤션에 따른 코드 수정

- 불필요한 self 생략
- 마크 주석을 통한 기능 분리
- 한줄짜리 메서드 구현 취소
- 파일명 변경
- Common 폴더 생성

yeeton37 committed:
feat: EntryTableViewCell 생성 및 오토레이아웃 적용

yeeton37 committed:
refactor: ExpoInformationViewController 내부 Label 수정

- 기존에 있던 visitorsLabel, locationLabel, durationLabel의 크기 변경

Jeon-Minsu committed:
feat: UITableViewDelegate 프로토콜을 채택하여 performSegue 메서드 구현

Jeon-Minsu committed:
refactor: EntryTableViewCell 오토레이아웃 수정

yeeton37 committed:
feat: label에 dynamic type 적용 및 주석 처리

yeeton37 committed:
refactor: 컨벤션 수정

Jeon-Minsu committed:
docs: README.md 업데이트

Jeon-Minsu committed:
Merge remote-tracking branch 'refs/remotes/origin/step3'

yeeton37 committed:
docs: README.md 수정

yeeton37 committed:
docs: README.md 최종 수정

leejun6694 commented on Expo1900/Expo1900/Common/AppDelegate.swift:
```diff
@@ -7,10 +7,21 @@
-
-
+final class AppDelegate: UIResponder, UIApplicationDelegate {
+    // MARK: Properties
```
👍

leejun6694 commented on Expo1900/Expo1900/Common/AppDelegate.swift:
```diff
@@ -7,10 +7,21 @@
+            return UIInterfaceOrientationMask.all
+        }
+        
+        return UIInterfaceOrientationMask.portrait
```
아래와 같이 구현할 수도 있습니다 : )
```
return shouldSupportAllOrientation == true
   ? .all
   : .portrait
```

Jeon-Minsu replied:
해당 방법으로 수정완료하였습니다!

yeeton37 replied:
해당 방법으로 수정했지만 쿠마가 아래 코멘트에서 알려주신 `supportedinterfaceorientations`프로퍼티를 사용한 방법으로 바꿔주어 위의 부분은 지워주게 되었습니다!

leejun6694 commented on Expo1900/Expo1900/Controller/EXPOInformationViewController.swift:
```diff
@@ -16,101 +17,122 @@ final class EXPOInformationViewController: UIViewController {
     @IBOutlet private weak var descriptionLabel: UILabel!
     @IBOutlet private weak var navigationButton: UIButton!
     
+    private let appDelegate = UIApplication.shared.delegate as! AppDelegate
```
강제 언래핑은 지양해주세요~!

Jeon-Minsu replied:
private let appDelegate = UIApplication.shared.delegate as? AppDelegate

위와 같이 수정하여, 강제 다운 캐스팅이 아닌 조건부 다운 캐스팅을 진행해보았습니다.

leejun6694 commented on Expo1900/Expo1900/View/EntryTableViewCell.swift:
```diff
@@ -0,0 +1,80 @@
+//
+import UIKit
+
+final class EntryTableViewCell: UITableViewCell {
```
💯

leejun6694 commented on Expo1900/Expo1900/View/EntryTableViewCell.swift:
```diff
@@ -0,0 +1,80 @@
+
+final class EntryTableViewCell: UITableViewCell {
+    // MARK: Properties
+    
```
`Exhibits value` 를 하나 전달 받아서, cell 안에서 세팅해주는건 어떨까요?
`prepareForReuse` 도 고민해봐야겠네요 : )

Jeon-Minsu replied:
- Exhibits value를 View에 넣어주기 위해서 exhibit 타입 프로퍼티를 생성하여 전달했습니다.
- 전달받은 exhibit을 통해 image, title, short_desc를 할당해주었는데, 셀이 재사용될 때마다 초기화를 해주지 않으면, 재사용되기 이전의 셀의 이미지와 텍스트가 그대로 다시 사용되는 것을 확인할 수 있었습니다.
- 이에 prepareForReuse 메서드에서 image, title, short_desc를 nil로 초기화를 해주었더니, 이번에는 스크롤을 위로 올리니 앞서 생성된 셀의 내용이 재사용으로 인해 Nil이 되버리고 말았습니다.
- 이를 위해서 CellForRowAt 메서드에서 해당 row의 koreaEntry가 nil이 아닌 값이면 image, title, short_desc를 다시 설정해주었습니다.
- 위의 과정을 다시 돌아보니, cellForRowAt에서 한번만 선언을 하면 되는 동작들이 너무 복잡하게 변한 것 같습니다.
- Exhibits value 를 하나 전달 받아서, cell 안에서 세팅해야 하는 이유가 무엇인지 궁금합니다.

- 코드(KoreaEntryViewController)
```swift
func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    let exhibit = koreaEntry[indexPath.row]
    
    EntryTableViewCell.exhibit = exhibit
    
    let cell = tableView.dequeueReusableCell(withIdentifier: "EntryTableViewCell", for: indexPath) as! EntryTableViewCell
    
    let entry = koreaEntry[indexPath.row]
    
    if entry != nil {
        cell.entryImageView.image = UIImage(named: koreaEntry[indexPath.row].imageName)?.withAlignmentRectInsets(UIEdgeInsets(top: -10, left: 0, bottom: -10, right: 0))
        cell.titleLabel.text = koreaEntry[indexPath.row].name
        cell.descriptionLabel.text = koreaEntry[indexPath.row].shortDescription

        cell.accessoryType = .disclosureIndicator
    }
    
    return cell
}
```

leejun6694 replied:
UI Component 는 private 으로 제한하고, 코드를 간단하게 할 수 있지 않을까 생각했습니다.
cell 의 높이가 계속 말썽인 것 같은데, 내일 같이 얘기해보시죠~!!

leejun6694 commented on Expo1900/Expo1900/Controller/EXPOInformationViewController.swift:
```diff
@@ -16,101 +17,122 @@ final class EXPOInformationViewController: UIViewController {
         super.viewWillAppear(animated)
         
-        self.navigationController?.isNavigationBarHidden = true
+        appDelegate.shouldSupportAllOrientation = false
```
https://developer.apple.com/documentation/uikit/uiviewcontroller/1621435-supportedinterfaceorientations
위 내용도 참고해볼까요?

yeeton37 replied:
이런 방법도 있었군요! 기존에 메서드 application(_:supportedInterfaceOrientationsFor:) 를 사용해서 ViewController에서 세로모드로 변경해주던 것을
ViewController에서 처음부터 프로퍼티 supportedInterfaceOrientations를 사용해서 방향을 변경해주는 방법으로 변경해주었습니다!
```swift

class navigationController: UINavigationController {
    override var supportedInterfaceOrientations: UIInterfaceOrientationMask {
        return self.topViewController?.supportedInterfaceOrientations ?? [.all]
    }
}

override var supportedInterfaceOrientations: UIInterfaceOrientationMask {
        return [.portrait]
    }
```

leejun6694 reviewed (changes requested):
수꿍, yeton 안녕하세요!
step 2 에서 같이 얘기했던 부분을 잘 적용해주셨네요 : )
step 3 리뷰 내용도 확인 부탁드립니다~!

leejun6694 commented:
> 1. UIApplication.shared.delegate as! AppDelegate 와 AppDelegate()의 차이

네, 목적에 맞는 singleton instance 를 사용할 것인지, 새로운 instance 를 만들 것인지의 차이입니다.
잘 이해하신 것 같네요~~!

> 2. 이미지도 dynamic type이 적용되야 하는지

이 부분은 좀 더 확인하고 코멘트 남기겠습니다 : )

> 3. ContentConfiguration과 Custom Cell

현재 구현된 코드에서 cell 의 높이를 설정해주는 부분이 있을까요?
table view 의 cell 높이를 설정하는 방법을 찾아볼까요??

> 4. 오토레이아웃 문제

3번이 먼저 해결되고 나서, 추후에 다시 확인해볼까요!?

> 5. Navigation Back Button Title 변경 문제

버그인가요...? 혹시 다른 캠퍼분들도 발생하는지 물어봐줄 수 있으실까요?
다만, `back button title` 을 명시하면 문제 없이 동작하네요..

Jeon-Minsu committed:
refactor: 코드 간결화 및 강제 언래핑 지양

- AppDelegate 파일 내 application 메서드 반환값을 삼항 연산자로 수정
- EXPOInformationViewController 파일 내 appDelegate 변수의 타입 캐스팅 방법을 조건부 다운 캐스팅으로 변경

yeeton37 committed:
refactor: supportedInterfaceOrientations 프로퍼티를 override해서 첫 번째 화면 세로모드로 고정

- 기존에 메서드 application(_:supportedInterfaceOrientationsFor:) 를 사용해서 ViewController에서 세로모드로 변경해주던 것을
프로퍼티 supportedInterfaceOrientations를 사용해서 방향을 변경해주는 방법으로 변경함

yeeton37 committed:
refactor: Cell의 프로퍼티를 KoreaEntryViewController에서 바꾸지 못하도록 메서드 setup을 생성해줌

- EntryaTableViewCell 내부에서 setup 메서드를 생성해서 기존에 KoreaEntryViewController에서 Cell의 이미지,타이틀,description을 변경해줬던 것을 Cell 내부에서만 변경할 수 있도록 해줌

yeeton37 commented:
> 2. ContentConfiguration과 Custom Cell

현재 구현된 코드에서 cell 의 높이를 설정해주는 부분이 있을까요?
table view 의 cell 높이를 설정하는 방법을 찾아볼까요??

-> Custom Cell을 만들어서 사용할 때는 cell의 높이를 
` entryImageView.heightAnchor.constraint(equalTo: heightAnchor).isActive = true`
와 같이 설정해주어서 해결이 됐습니다.
위에서 질문드렸듯이, cell 자체에 heightAnchor를 설정해주는 것과, cell의 contentView에 
heightAnchor을 설정해주는 것의 차이가 무엇인지 궁금합니다.

> 5. Navigation Back Button Title 변경 문제

버그인가요...? 혹시 다른 캠퍼분들도 발생하는지 물어봐줄 수 있으실까요?
다만, back button title 을 명시하면 문제 없이 동작하네요..

-> 다른 캠퍼분들도 똑같은 문제가 발생하는 것 같습니다. Accessibility를 통해 Dynamic Type의 크기를 키우면 back button title에다가 명시해도 결국 title의 크기가 커져서 똑같이 back으로 변경되는 것을 확인했습니다...ㅜㅜ 그래서 원래 이렇게 되는 것이 맞는 것인지, 아니면 해결을 해주어야 하는 부분인지... 잘 모르겠습니다


위의 질문에서 3,5번에 대해서는 쿠마와 이야기 나눠보고싶습니다!!  나머지 코멘트 달아주신 부분은 모두 해결했습니다!

leejun6694 reviewed (approved):
리뷰 내용을 잘 반영해주셨습니다 : )
미흡한 부분이 있는데, 내일 같이 얘기해보아요~!
